File: src/EnumeratorUtils/Library.cs
using System.Collections;
using System.Collections.Generic;

namespace EnumeratorUtils {
    public static class Library {
        public static IEnumerable<T> Cast<T>(this IEnumerator data) {
            while (data.MoveNext()) {
                yield return (T)data.Current;
            }
        }
    }
}

File: src/FSharp/Library.fs
namespace EnumeratorUtils

open System.Collections
open System.Runtime.CompilerServices

(*
// Not visible from C#

type IEnumerator with
    member this.Cast<'a>() =
        seq {
            while this.MoveNext() do yield this.Current
        }
        |> Seq.cast<'a>
*)

[<Extension>]
type Extensions =
    [<Extension>]
    static member Cast<'a>(datas: IEnumerator) =
        seq {
            while datas.MoveNext() do yield datas.Current
        } |> Seq.cast<'a>